src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -122,7 +123,7 @@ static ArrowErrorCode ArrowDeviceCudaAllocateBuffer(struct ArrowDevice* device,
   switch (device->device_type) {
     case ARROW_DEVICE_CUDA: {
       CUdeviceptr dptr = 0;
-      err = cuMemAlloc(&dptr, (size_t)size_bytes);
+      err = cuMemAllocAsync(&dptr, (size_t)size_bytes, *stream);
```

Comment:
It's probably worth checking if the finding in https://github.com/apache/arrow-nanoarrow/pull/534 is relevant here. `cuMemAlloc()` is documented not to like `0` input for the `byte_size` parameter, but `cuMemAllocAsync()` has no such limitation AFAI can see.